norse/torch/functional/stdp.py

```diff
@@ -1,60 +1,407 @@
+    """
+    # ReLU to ensure weights outside [w_min, w_max] move towards it (when wrongly initialized)
+    return torch.nn.functional.relu(w_max - w) * eta_plus
+
+
+def _A_plus_hard(w, w_max, eta_plus):
+    """Hard upper bound for multiplicative STDP
 
-def A_plus_hard(w, w_max, eta_plus):
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        w_max (float): upper bound on synaptic weight (should be > w_min)
+        eta_plus (float): contribution of pre-post activity to synaptic weight (should be >= 0)
+    """
     return heaviside(w_max - w) * eta_plus
 
 
-def A_minus_soft(w, eta_minus):
-    return w * eta_minus
+def _A_minus_soft(w, w_min, eta_minus):
+    """Soft lower bound for multiplicative STDP
+
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        w_min (float): lower bound on synaptic weight (should be < w_max)
+        eta_minus (float): contribution of post-pre activity to synaptic weight (should be >= 0)
+    """
+    # ReLU to ensure weights outside [w_min, w_max] move towards it (when wrongly initialized)
+    return torch.nn.functional.relu(w - w_min) * eta_minus
+
+
+def _A_minus_hard(w, w_min, eta_minus):
+    """Hard lower bound for multiplicative STDP
+
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        w_min (float): lower bound on synaptic weight (should be < w_max)
+        eta_minus (float): contribution of post-pre activity to synaptic weight (should be >= 0)
+    """
+    return heaviside(w - w_min) * eta_minus
+
+
+def linear_soft_multiplicative_stdp_step(
+    w: torch.Tensor,
+    z_pre: torch.Tensor,
+    z_post: torch.Tensor,
+    state: STDPState,
+    p: STDPParameters = STDPParameters(),
+    dt: float = 0.001,
+) -> Tuple[torch.Tensor, STDPState]:
+    """Soft-bound multiplicative STDP step for nn.Linear layer, meaning change in
+    synaptic weight is dependent on current synaptic weight. Based on online implementation
+    from http://www.scholarpedia.org/article/Spike-timing_dependent_plasticity
+
+    Parameters:
+        w (torch.Tensor): synaptic weight
+        z_pre (torch.Tensor): presynaptic spikes
+        z_post (torch.Tensor): postsynaptic spikes
+        state (STDPState): STDP state
+        p (STDPParameters): STDP parameters
+        dt (float): integration time step
+    """
+    dx = dt * p.tau_pre_inv * (-state.x + p.a_pre * z_pre)
+    x_decayed = state.x + dx
+
+    dy = dt * p.tau_post_inv * (-state.y + p.a_post * z_post)
+    y_decayed = state.y + dy
+
+    # w has shape (post, pre), so spikes/trace for postsynaptic neuron have to be transposed
+    # Pre- and postsynaptic neurons can be any 2-4D (BCHW) shape
+    dw_plus = _A_plus_soft(w, p.w_max, p.eta_plus) * x_decayed * z_post[..., None]
+    dw_minus = _A_minus_soft(w, p.w_min, p.eta_minus) * y_decayed[..., None] * z_pre
+
+    # TODO: what to do with batch dimension in an online rule? Sum? Nothing?
```

Comment:
Since you're not explicitly changing the weights, I think this is a sound approach. If batch normalisation or similar is required, it's entirely possible to implement that afterwards. 
It would also be a possibility to just expose al batches as-is. Would that be desirable in any situation?

Reply:
I'm not sure whether that would be desirable.. However, there might be multiple options for combining batches (mean, sum), so it would be most flexible to leave this choice to the user. Both BindsNET and PySNN do some sort of aggregation though I think

Reply:
Hmm I always found the trade-off between freedom and usability quite hard. Do you have any opinions @cpehle?

Reply:
One option would be to pass in a function as an argument. It also should be possible to integrate weight changes over time (relative to the integration timestep for the neuron dynamics). So it might be necessary to carry though the dw_+, dw_- changes over multiple time steps.

Reply:
It seems like this isn't exactly a red lamp in the sense that it'll become an immediate problem. Would you mind if I resolved this conversation for now and then we can always add the feature later, if needed?